### PR TITLE
[REF TEST] refactor and add unit tests to DWI EPI pipeline utility functions

### PR DESCRIPTION
**Pull request by NicolasGensollen**
- refactor `change_itk_transform_type` to use more modern `pathlib` functionalities
- rename confusing `expend_matrix_list(in_matrix, in_bvec)` to `broadcast_matrix_filename_to_match_b_vector_length(matrix_filename: str, b_vectors_filename: str) -> list` and refactor
- rename `rotate_bvecs(in_bvec, in_matrix)` to `rotate_b_vectors(b_vectors_filename: str, matrix_filenames: list) -> str`
- Add type hints and documentation for the three functions
- Add unit tests for the three functions

### CODE REVIEW

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py [hunk 1]
```diff
@@ -98,42 +109,59 @@ def ants_warp_image_multi_transform(fix_image, moving_image, ants_warp_affine):
     return out_warp
 
 
-def rotate_bvecs(in_bvec, in_matrix):
-    """Rotate the input bvec file accordingly with a list of matrices.
+def rotate_b_vectors(b_vectors_filename: str, matrix_filenames: list) -> str:
+    """Rotate the B-vectors contained in the input b_vectors_filename file
```
Comment (MatthieuJoulot) [inline clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:113]:
```suggestion
    """Rotates the B-vectors contained in the input b_vectors_filename file
```

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py [hunk 2]
```diff
@@ -42,46 +42,57 @@ def rename_into_caps(
...
 
-    updated_affine_file = os.path.join(os.getcwd(), "updated_affine.txt")
 
-    with open(updated_affine_file, "wt") as f:
-        for line in new_file_lines:
-            f.write(line)
+def broadcast_matrix_filename_to_match_b_vector_length(
+    matrix_filename: str, b_vectors_filename: str
+) -> list:
+    """Return a list of the matrix filename repeated as many times as there are B-vectors."""
```
Comment (MatthieuJoulot) [inline clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:81]:
```suggestion
    """Returns a list of the matrix filename repeated as many times as there are B-vectors."""
```

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py [hunk 3]
```diff
@@ -42,46 +42,57 @@ def rename_into_caps(
...
+    return broadcast_filename_into_list(matrix_filename, len(b_vectors))
 
-    bvecs = np.loadtxt(in_bvec).T
-    out_matrix_list = [in_matrix]
 
-    out_matrix_list = out_matrix_list * len(bvecs)
-
-    return out_matrix_list
+def broadcast_filename_into_list(filename: str, desired_length: int) -> list:
+    """Return a list of provided filename repeated to match the desired length."""
```
Comment (MatthieuJoulot) [inline clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:94]:
```suggestion
    """Returns a list of provided filename repeated to match the desired length."""
```

**Review by MatthieuJoulot — CHANGES_REQUESTED** [review]
Looking good ! Just a few minor changes and a few test factorization :)

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py (hunk 1 above)
Comment (NicolasGensollen) [reply clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:113]: Actually, imperative style is usually considered more Pythonic (see [PEP-257](https://peps.python.org/pep-0257/) for more information).

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py (hunk 2 above)
Comment (NicolasGensollen) [reply clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:81]: same

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py [hunk 4]
```diff
@@ -42,46 +42,57 @@ def rename_into_caps(
     )
 
 
-def change_itk_transform_type(input_affine_file):
+def change_itk_transform_type(input_affine_file: str) -> str:
     """Change ITK transform type.
```
[the remark this answers is not in the document]
Comment (NicolasGensollen) [reply clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:46]: same

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py (hunk 3 above)
Comment (NicolasGensollen) [reply clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:94]: same

File: test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py [hunk 5]
```diff
@@ -21,6 +23,152 @@ def test_extract_sub_ses_folder_name():
     )
 
 
+def test_change_itk_transform_type_error(tmp_path):
+    with pytest.raises(FileNotFoundError):
+        change_itk_transform_type(str(tmp_path / "affine.txt"))
+
+
+def test_change_itk_transform_type_empty_file(tmp_path):
```
[the remark this answers is not in the document]
Comment (NicolasGensollen) [reply test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py:31]: I don't understand, I'm not doing any iteration in this test. The tested function is called once.

File: test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py [hunk 6]
```diff
@@ -21,6 +23,152 @@ def test_extract_sub_ses_folder_name():
...
+    (tmp_path / "affine.txt").touch()
+
+    assert change_itk_transform_type(str(tmp_path / "affine.txt")) == str(
+        tmp_path / "updated_affine.txt"
+    )
+    assert (tmp_path / "updated_affine.txt").read_text() == ""
+
+
+def test_change_itk_transform_type_no_transform_line(tmp_path):
```
[the remark this answers is not in the document]
Comment (NicolasGensollen) [reply test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py:40]: same answer

File: test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py [hunk 7]
```diff
@@ -21,6 +23,152 @@ def test_extract_sub_ses_folder_name():
...
+    )
+    origin_affine.write_text(origin_content)
+
+    assert change_itk_transform_type(str(origin_affine)) == str(
+        tmp_path / "updated_affine.txt"
+    )
+    assert (tmp_path / "updated_affine.txt").read_text() == origin_content
+
+
+def test_change_itk_transform_type(tmp_path):
```
[the remark this answers is not in the document]
Comment (NicolasGensollen) [reply test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py:53]: same answer

File: clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py (hunk 1 above)
Comment (MatthieuJoulot) [reply clinica/pipelines/dwi_preprocessing_using_t1/dwi_preprocessing_using_t1_utils.py:113]: Alright ! Feels strange and unnatural, but it is indeed what they say !

File: test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MatthieuJoulot) [reply test/unittests/pipelines/dwi_preprocessing_using_t1/test_dwi_preprocessing_using_t1_utils.py:31]: ha yes sorry you are right

**Review by MatthieuJoulot — APPROVED** [review]
All good :)

**Pull request merged**